### ISSUE TITLE
method iso8601() is wrong to be strictly

### ISSUE THREAD

**Issue by worthmine** [houseabsolute/DateTime.pm#109]
Hi, I made [an Issue](https://github.com/bk/Data-ULID/issues/3) on [Data-ULID](https://github.com/bk/Data-ULID).

bk, Author of this module said:
> yes, there is a bit of a problem here – but it consists in an inconvenience in formatting DateTime objects,
I really agree with this closing.

The core matter is not in this module; In DateTime.

Even if named `iso8601()` is, there is **no method based on ISO 8601 strictly**.see [RFC 3339](https://tools.ietf.org/html/rfc3339#section-5.8) 

ISO 8601 defines that timestamps require 'Z' on end of strings if timezone was 'UTC', else they require the offset like "+09:00"
if there is no "Z" we (the human and machines) have to parse that is with a localtime. it's wrong.

So soon, at the maybe until  next weekend, I will send a PR to fix it.

Before that,  please tell me the reasons why it is no necessary if it is.

**Comment by autarch** [comment]
At this point, wrong or not, I don't think it can be fixed. There's too much chance of breaking someone's code with changes like this, unfortunately.

**Comment by worthmine** [comment]
thank you for fast replying!

if I *change* the method `iso8601()` it will be. but if I *add* the method that is correct, then what will be happened on?

**Comment by autarch** [comment]
Yes, a new method is a good idea. How about we call it `iso8601_with_offset`?

**Comment by autarch** [comment]
However, I should point out that this won't necessarily fix your original issue. The default stringification of a `DateTime` will still just call `$self->iso8601`.

**Comment by worthmine** [comment]
> How about we call it iso8601_with_offset?

I agree!

And thanks for pointing out.
I don't stick to stringify with overloading.
Just I need a method that is quite according to ISO 8601. because that is not yet.

btw, overloading with `q{"}` calls `stringify` with no argument.

and `stringify` defined like below:
```
 sub stringify {
    my $self = shift;

    return $self->iso8601 unless $self->{formatter};
    return $self->{formatter}->format_datetime($self);
}
```

Therefore I think like this:
    if we changed [this](https://github.com/houseabsolute/DateTime.pm/blob/753f9019fab4f04ef7afc266baee5d2109f54ed9/lib/DateTime.pm#L937) to 
```diff
-return $self->iso8601 unless $self->{formatter};
+return $self->iso8601() unless $self->{formatter};
```
   there will be nothing to be happened.

So does iso8601, it is just [defined as](https://github.com/houseabsolute/DateTime.pm/blob/753f9019fab4f04ef7afc266baee5d2109f54ed9/lib/DateTime.pm#L956):
```
sub iso8601 { $_[0]->datetime('T') }
```
so probably it is **NOT** affected by it own arguments.

It's my 2 cents, *adding* is better idea to avoid other issues but *changing* is also not too bad. 

Let's start by changing it for now.

**Comment by n1vux** [comment]
Would we consider an option (settable by constructor, setter, or ENVironment?) to use strict iso8601 version, what ever it's called, for `stringify` ?

**Comment by worthmine** [comment]
@n1vux sounds good

**Comment by autarch** [comment]
> Would we consider an option (settable by constructor, setter, or ENVironment?) to use strict iso8601 version, what ever it's called, for `stringify` ?

The `formatter` option to the constructor already lets you override the default.

**Comment by autarch** [comment]
So FWIW I took a look at the [only ISO8601 spec I could find online (2004)](http://dotat.at/tmp/ISO_8601-2004_E.pdf). It doesn't _require_ the time zone to be included in a datetime format. It's simply an option.

RFC3339 specifies a very strict, very small subset of ISO8601.

**Comment by n1vux** [comment]
Yes, a subset or profile for use in internet protocols - where the assumption of a single common presumptive TZ (floating) is invalid.

And it's not even a proper subset; a TZ offset of -0:00 isn't officially allowed per ISO8601. (Unfortunate!)

So the stricter version for safe internet proto use might be better named `$dt->rfc_3339()` ?

**Comment by autarch** [comment]
Yeah, naming it `rfc3339` is probably best

**Comment by worthmine** [comment]
I'm sorry to be late.
I have a time to make PR today and agree with naming it `rfc3339()`.
How about underbar? 
I think it's enough to omit.

**Comment by autarch** [comment]
Thanks, but I already did it - https://github.com/houseabsolute/DateTime.pm/pull/113

**Comment by worthmine** [comment]
I see. 
it seems I was too late :(

**Comment by autarch** [comment]
Heh, sorry about that. But please do review that PR.

**Comment by worthmine** [comment]
Of course I do.

**Comment by worthmine** [comment]
I approved the review but status didn't change. What should I do?

**Comment by autarch** [comment]
> I approved the review but status didn't change. What should I do?

I don't think you can be a reviewer in GitHub terms because you don't have write access to the repo.

**Comment by worthmine** [comment]
Does GitHub require anybody else me who has write access to the repo?

I thought I can review because you asked me to do that.

**Comment by autarch** [comment]
@worthmine Let me clarify. I wanted you to take a look at the code to see if it did what you want. It seems like you did that, which is great.

But you can't participate in a full GitHub Pull Request Review because you are not part of the repo.

**Comment by worthmine** [comment]
Thank you for explaining.

If I wrote the PR then there is another way. it's caused that I'm too late and didn't write it by myself, isn't it?

**Comment by autarch** [comment]
If you'd made the PR then _I_ would be reviewing _your_ work. In this case, I've asked other people with access to the repo to do the PR review. It's just a way to double check my work.

I try to do this for changes to DateTime because it's so widely used in the Perl world.

**Comment by worthmine** [comment]
Fine, I see.
At least functionally, there is nothing I can mention for this issue anymore, isn’t it?
 I satisfy because my wish came true 😁

**Comment by worthmine** [comment]
It's my last wish.
Please write my name as 'worthmine' instead of my full name in Changes.

**Comment by autarch** [comment]
> It's my last wish.
> Please write my name as 'worthmine' instead of my full name in Changes.

Done.

**Comment by autarch** [comment]
Resolved in v1.53
